Add self evaluation on Metal3.io's technical updates for Graduation (#53)
metal3-io/community: self-evaluation/2026-metal3-graduation-self-technical-review.md

     not require any commercial cloud provider service.
- Describe the intended types of organizations who would benefit
  from adopting this project.
   - Telco operators (Ericsson, Deutsche Telekom), enterprise
     infrastructure providers (Red Hat, SUSE, Fujitsu), telco
     cloud infrastructure platforms (Sylva, Airship), AI/ML
     infrastructure (Mistral AI), and large-scale on-premise
     environments (Mirantis). Deployments range from
     single-server setups to 2000+ node production
     environments.
- Please describe any completed end user research and link to any
  reports.
   - Adopter feedback is gathered through the CNCF incubation
     adopter interviews (Ericsson, Red Hat, SUSE, Deutsche
     Telekom, and a fifth Telco adopter — see
     [incubation DD][incubation]), direct engagement at community
     meetings, KubeCon presentations, and the
     [Metal3 annual meetup][meetup-notes].

### Usability

- How should the target personas interact with your project?
   - Operators interact entirely through Kubernetes-native APIs
     by creating and managing BaremetalHost, Metal3Machine,
     Metal3Cluster, IPPool, and IPClaim custom resources using

>      standard tools like kubectl and clusterctl.

Reviewer: Unsure if this point should be about interacting with the project/community itself or with the software that we produce. Is there any previous example we could check?

Author: yes, check this for example which I did for k8gb and it was reviewed, I had the same understanding as you but then the Emily from TOC corrected me https://github.com/cncf/toc/pull/2029#discussion_r2789095524

- Describe the user experience (UX) and user interface (UI) of
  the project.
   - Metal3 has no graphical UI; the interface is the Kubernetes

>      API with CRDs and status conditions. Operators use kubectl

Reviewer: ```suggestion
     API with CRDs and status conditions. Operators use kubectl (or other Kubernetes API clients)
```

     to inspect BaremetalHost status, events, and conditions,
     which surface the full hardware lifecycle state machine
     (registering → inspecting → available → provisioning →
     provisioned → deprovisioning).

>    - The core experience is the BaremetalHost Custom Resource —

Reviewer: CAPI was highlighted in the primary use-case above, so I think we should add one point here also that includes it. I.e. describe how it is used with CAPI (not how we integrate, which is the next point below)

   - Architecture documented at
     [book.metal3.io/architecture][arch].
- Define any specific service dependencies the project relies on
  in the cluster.
   - Kubernetes API server (with etcd), Ironic deployed via
     the [Ironic Standalone Operator][irso] (recommended for
     production), and a MariaDB or SQLite backend for Ironic
     state. PXE boot requires DHCP/TFTP infrastructure unless
     virtual media boot is used. No external database or SaaS
     dependency is required.
- Describe how the project implements Identity and Access
  Management.
   - Metal3 leverages Kubernetes-native security primitives
     rather than building its own IAM system. It follows the
     principle of least privilege, separating controller
     permissions per component (BMO manages hosts, CAPM3
     manages metal3 machines, IPAM manages addresses).
   - Each operator runs with its own Kubernetes ServiceAccount
     with RBAC scoped to minimal required permissions. BMC
     credentials are stored in Kubernetes Secrets referenced by
     BaremetalHost resources and are never logged or exposed in
     status fields.
   - Developers only need permission to create/modify the

>      BaremetalHost CRD in their specific namespace.

Reviewer: I guess we don't need to mention all the CRDs, but let's at least acknowledge that there are more than one.
```suggestion
     BaremetalHost and other CRDs in their specific namespace.
```

- Describe how the project has addressed sovereignty.
   - Metal3 runs entirely on-premise with no external cloud
     dependencies. It uses standard BMC protocols (Redfish,
     IPMI) and is not tied to any specific hardware vendor.
     All data — including BMC credentials, host inventory,
     and provisioning state — remains within the operator's
     Kubernetes cluster. The project supports air-gapped

>      deployments.

Reviewer: Here I am not sure if the point is about the project/community or the software/architecture. The latter is well described but the first is missing, if it is needed.

Author: AFAIK this is talking about the later, from the same k8gb review experience.

- Describe any compliance requirements addressed by the project.
   - The project has earned an
     [OpenSSF Best Practices badge][ossf], uses DCO for all
     commits, and follows [CNCF SSCP][sscp] best practices
     with signed releases and SBOMs. The on-premise,
     air-gap-capable deployment model supports adopters' own
     compliance requirements (e.g., Telco security standards,
     data sovereignty regulations).
- Describe the project's High Availability requirements.
   - BMO and CAPM3 controllers support leader election via
     controller-runtime for HA deployments. Ironic can be
     deployed in active-passive HA configurations using the
     [Ironic Standalone Operator][irso]. The BaremetalHost state
     machine is idempotent and resumes correctly after controller
     restarts.
- Describe the project's resource requirements, including CPU,
  Network and Memory.
   - Pretty lightweight: BMO and CAPM3 controllers each require

>      approximately 256Mi–512Mi memory and 0.5 CPU in typical

Reviewer: Do we have any source for this? I don't doubt it, although I would have expected less. Hard to say what is typical of course.

Author: You are right, these were ballpark numbers and don't have any source, I will remove the numbers to avoid confusion

     deployments. Resource usage scales linearly with the
     number of managed BaremetalHosts. The Ironic Python
     Agent (IPA), which runs temporarily on each host during
     provisioning and inspection, is the main per-host
     resource consumer. Ironic itself can be tuned for
     large-scale deployments.
- Describe the project's storage requirements, including its use
  of ephemeral and/or persistent storage.
   - Ironic requires persistent storage for its database
     (typically 1–10Gi depending on fleet size). IPA images

>      require HTTP-accessible storage (~500Mi). Controllers

Reviewer: Is it normal to have several Gi Ironic database? 😮 
I think it would be good to mention also disk image storage for "production workload", not just IPA.

Author: Same, I think I found it in some openstack docs, but I will remove any number claim here since I dont find it anymore

     themselves are stateless — all state lives in Kubernetes
     CRDs.
- Please outline the project's API Design:
   - Describe the project's API topology and conventions
      - Metal3 uses Kubernetes CRDs following standard API
        conventions: BaremetalHost,
        Metal3Machine, Metal3Cluster, Metal3MachineTemplate,

>         IPPool, IPClaim, and IPAddress.

Reviewer: This is missing the topology part of it. We list the main CRDs and conventions, but we have nothing about how they relate. I think we need to add a bit more, at minimum that the Metal3* objects tie into CAPI. Probably it would be good to mention also how they are grouped into controllers.

   - Describe the project defaults
      - The recommended Ironic deployment overlay enables basic
        authentication and TLS. Newly created BaremetalHosts
        go through registering → inspecting → available by
        default. Inspection is enabled by default for new

>         hosts.

Reviewer: Please add something about CAPM3, IPAM and BMO. I think we could explain the defaults that come from deploying CAPM3 and IPAM using `clusterctl`. E.g. controller defaults, certificates and webhooks. Maybe also something about RBAC that comes by default with the manifests?

   - Outline any additional configurations from default to make
     reasonable use of the project
      - You'll typically configure: BMC credentials per host,
        provisioning image URL, root device selection hints,
        network data for the host, and RAID/BIOS configuration.
        The [quick start][quickstart] walks through common

>         setups.

Reviewer: I am not sure about this. It seems like normal part of using Metal3, just like creating the BMHs. I was expecting here something like Ironic network configuration (through IRSO) and maybe setting up image server. These are things we cannot have proper defaults for because they differ between environments.

   - Describe any new or changed API types and calls — including
     to cloud providers — that will result from this project
     being enabled and used

>       - Adds BaremetalHost, Metal3Machine, Metal3Cluster,

Reviewer: We have quite a lot more CRDs but I am not sure how detailed this should be. What about HostFirmwareComponents, HardwareData, Metal3Data(Template), also Ironic (from IRSO) and most recently HostClaim.

        Go module dependencies. Container base images are
        regularly updated. CI runs linters and vulnerability
        scanners on every PR. Published container images include
        SBOMs. The project has earned an
        [OpenSSF Best Practices badge][ossf].
   - Describe how the project has evaluated which features will
     be a security risk to users if they are not maintained by
     the project?
      - The main security-sensitive areas are BMC credential
        storage and access, Ironic API exposure, container image
        provenance, and network-level access to provisioning
        infrastructure — all mentioned in the
        [security self-assessment][self-assessment].
- Cloud Native Threat Modeling
   - Explain the least minimal privileges required by the project
     and reasons for additional privileges.
      - BMO needs read/write on BaremetalHost CRDs, read on
        Secrets (for BMC credentials), and write on Events.
        CAPM3 needs read/write on Metal3Machine/Cluster CRDs
        and read on BaremetalHost. IPAM needs read/write on
        IPPool/IPClaim/IPAddress CRDs. RBAC permissions are

>         defined in each controller's kustomize manifests.

Reviewer: Maybe we could simplify this by saying that each controller needs read/write access to the CRDs they reconcile. Then some of them also need access to secrets, configmaps for configuration. CAPM3 is a bit special since it also interacts with CAPI resources. This could be pointed out separately. IRSO is also interesting since it handles Kubernetes built-in resources (Deployment, Pod, etc).

To address the question about additional privileges, perhaps we could mention that at least BMO can run in a namespaced mode. The default is still cluster wide access, so this is "additional privileges" that we grant for ease of use.

   - Describe how the project is handling certificate rotation
     and mitigates any issues with certificates.
      - TLS certificates for Ironic communication can be managed
        via cert-manager or manual rotation. The Ironic
        Standalone Operator supports automated certificate
        management. Controller webhook certificates are managed

>         by controller-runtime's built-in certificate handling.

Reviewer: I think we should mention that these certificates are managed by cert-manager. Cert-manager handles rotation of them and provides the secrets that are then consumed by the controllers.

     secure software supply chain best practices.
      - Released container images are signed with keyless
        cosign and include SPDX SBOM attestations in the
        release workflows. Go module dependencies are pinned in
        `go.sum` and verified by Go module checksum
        validation. Dependabot and periodic OSV scanning are
        used to detect known vulnerabilities. The project
        follows [CNCF SSCP][sscp] best practices.

## Day 1 - Installation and Deployment Phase

### Project Installation and Configuration

- Describe what project installation and configuration look like.
   - Install via `clusterctl` or kustomize with your BMC
     network and provisioning interface configured, then deploy
     Ironic containers. Configuration is driven by Kubernetes
     ConfigMaps, environment variables, and CRD spec fields on
     BaremetalHost and Metal3Cluster resources.
     [Quick start here][quickstart] and
    [metal3-dev-env][dev-env] for local reproduction.
    Testing investment is primarily in Cluster API
    test-framework-based e2e coverage across the component
    repositories, with metal3-dev-env retained as a secondary

>     validation path.

Reviewer: I would skip this part. Development and CI is not the same as installation and configuration.
Instead, we could maybe describe what bootstrapping looks like?

> Create a bootstrap cluster first (unless the user already has a production cluster they wish to use). For example kind or minikube can be used for this. Make sure to also install cert-manager.
> Deploy Ironic using IRSO, if it is not already setup separately (e.g. as part of OpenStack).
> Deploy Bare Metal Operator by applying the release manifests.
> (Optional) Initialize CAPI and providers using `clusterctl` or the CAPI operator.
> The cluster is now ready for registering BareMetalHosts. If this is just a bootstrap cluster, it can be used to create the persistent management cluster at this point. Once that cluster is operational, use `clusterctl move` to make it the management cluster. See the CAPI docs for more details.

### Project Enablement and Rollback

- How can this project be enabled or disabled in a live cluster?
  Please describe any downtime required of the control plane or
  nodes.
   - Simple deployment apply/delete. BMO, CAPM3, and IPAM are
     standard Kubernetes Deployments — no control plane
     downtime. Existing bare metal hosts continue running
     unaffected when controllers are removed.
- Describe how enabling the project changes any default behavior
  of the cluster or running workloads.
   - Nothing changes until you create a BaremetalHost or
     Metal3Machine resource. Enabling Metal3 installs CRDs and

>      starts controllers that watch for these resources.

Reviewer: ```suggestion
   - Enabling Metal3 does not affect the clusters default behavior or running workload, other than by consuming some resources.
```

- Describe how the project tests enablement and disablement.
   - CI covers this in multiple layers: BMO e2e tests and
     Cluster API test-framework-based integration/feature/upgrade
     tests across CAPM3 and IPAM repositories. Those jobs
     exercise install → provision → deprovision → uninstall,
     including CRD cleanup and controller removal across
     supported release branches. metal3-dev-env scenarios are
     kept as secondary validation.
- How does the project clean up any resources created, including
  CRDs?
   - Controllers use finalizers on BaremetalHost and
     Metal3Machine resources to ensure hardware is properly
     deprovisioned before deletion. CRDs can be removed after
     all CR instances are deleted, and `clusterctl delete

>      --infrastructure metal3` handles full cleanup.

Reviewer: ```suggestion
     --infrastructure metal3 --ipam metal3` handles full cleanup of CAPM3 and IPAM. BMO and Ironic can be removed the same way they were installed, e.g. by deleting the release manifests for BMO and the Ironic object for Ironic.
```

   - Rollback is performed by redeploying the previous
     controller version; the CRD state machine is designed to
     be forward-compatible, and controllers reconcile from
     current resource state regardless of version.
     **Ironic database migrations are forward-only and
     represent the main rollback risk** — operators should
     snapshot the Ironic database before upgrading.
- How can a rollout or rollback fail? Describe any impact to
  already running workloads.
   - CRD schema incompatibilities with stored resources, or new
     controller version introducing a bug in the state machine.
     Ironic database migrations being forward-only mean
     downgrades may require database restoration.
     Already-provisioned bare metal hosts continue running
     unaffected — only lifecycle operations are affected.
- Describe any specific metrics that should inform a rollback.
   - Watch for: BaremetalHost resources stuck in `error` state,
     increased controller restart count, or elevated
     reconciliation error rates in Prometheus metrics.
- Explain how upgrades and rollbacks were tested and how the
  upgrade->downgrade->upgrade path was tested.

>    - Upgrade tests are part of the CI matrix, testing rolling

Reviewer: I think the only upgrade->downgrade->upgrade path we test is really webhook conversions. All other tests are only plain upgrades, right? It would be good to check what CAPI has to say about this. Is there proper support for downgrade or why is there no test for it?

Author: CAPI doesn't support downgrades , I will change the wording here if this needs more clarity on that

   - Alpha features are gated behind feature flags in controller
     configuration and are disabled by default. API versions
     follow Kubernetes conventions (v1alpha1, v1beta1).
     Component maturity:

     | Component | Maturity | Version |
    | :---------- | :--------- | :-------- |
      | Baremetal Operator (BMO) | **Stable** | v0.13.x |
      | Cluster API Provider Metal3 (CAPM3) | **Stable** | v1.13.x |
      | IP Address Manager (IPAM) | **Stable** | v1.13.x |
     | Ironic Standalone Operator (IrSO) | **Beta** | v0.x |
     | Ironic Image | **Stable** | Released with BMO |
     | IPA Downloader | **Stable** | Released with BMO |
     | MariaDB Image | **Stable** | Released independently |

## Day 2 - Day-to-Day Operations Phase

### Scalability/Reliability

- Describe how the project increases the size or count of
  existing API objects.
   - Object count scales primarily with managed hosts: each host
     maps to one BaremetalHost plus associated Secret and
     optional IPClaim/IPAddress resources. CI includes
     scalability scenarios that exercise increased API-object
     counts and reconciliation behavior at higher host counts,

>      with no fan-out amplification pattern in controller logic.

Reviewer: I think the question here is about existing API objects, such as pods and services and configmaps. So we should focus on those.

```suggestion
   - Each controller adds a few CRDs, a Deployment and Service, RBAC and webhooks. The controllers also emit Events, mostly based on user activity. The main source of API object count increase comes from user workloads. Each provisioned BareMetalHost may be associated with a number of Secrets.
     CI includes
     scalability scenarios that exercise increased API-object
     counts and reconciliation behavior at higher host counts,
     with no fan-out amplification pattern in controller logic.
```

Author: humm, I still think this also talks about the API objects we create as part of the project but anyhow I will change the wording to incorporate your input as well

     queues, exponential backoff, rate limiting, and leader
     election. The BaremetalHost state machine is idempotent —
     controllers reconcile from observed state and recover
     automatically from transient failures.

### Observability Requirements

- Describe the signals the project is using or producing,
  including logs, metrics, profiles and traces. Please include
  supported formats, recommended configurations and data
  storage.
   - Produces Prometheus metrics (reconciliation duration, error
     counts, queue depth) via controller-runtime's built-in
     endpoint on `:8443/metrics`. Controller logs are emitted
     through controller-runtime logging (klog/logr, with zap in
     some components/configurations). Ironic services emit their
     own service logs. Distributed tracing is not enabled by
     default.
- Describe how the project captures audit logging.
   - Relies on Kubernetes audit logging for API operations. The
     controller logs all reconciliation actions. BMC credential
     access is logged but credentials are redacted from all log
     output.
- Describe any dashboards the project uses or implements as well
  as any dashboard requirements.

>    - Metal3 does not ship built-in Grafana dashboards; however,

Reviewer: This is focusing on Grafana only. I think we should either just say that we do not ship any dashboards or we should mention also things like Headlamp.

     the Prometheus metrics are scrapable and adopters build
     dashboards for BaremetalHost state distribution,
     provisioning latency, and error rates.
- Describe how the project surfaces project resource requirements
  for adopters to monitor cloud and infrastructure costs, e.g.
  FinOps

>    - BaremetalHost status exposes hardware inventory (CPU, RAM,

Reviewer: ```suggestion
   - BareMetalHost status exposes hardware inventory (CPU, RAM,
```

     disk, NIC details) discovered during inspection, enabling
     adopters to build asset tracking and capacity planning on
     top of the CRD data.
- Which parameters is the project covering to ensure the health
  of the application/service and its workloads?
   - Health is tracked across BMH, CAPM3, and IPAM resources:
     BaremetalHost `.status.conditions` and
     `.status.operationalStatus`, Metal3Machine/Metal3Cluster
     readiness and error conditions, and IPPool/IPClaim/
     IPAddress allocation state. Controller health is exposed

>      via `/healthz` and `/readyz` endpoints.

Reviewer: ```suggestion
     via `/healthz` and `/readyz` endpoints, and monitored via liveness and readiness probes.
```

     Metal3Machine/Metal3Cluster resources should converge to
     ready states, IPAM objects should show successful claim and
     allocation, reconciliation metrics should be healthy, and
     no persistent errors should appear in controller logs.
- Describe the SLOs (Service Level Objectives) for this project.
   - Not formally defined by the project. Adopters define SLOs
     based on their fleet size and operational requirements.
- What are the SLIs (Service Level Indicators) an operator can
  use to determine the health of the service?
   - Reconciliation success rate, provisioning success rate,
     BaremetalHost time-to-available, Metal3Machine/Cluster
     reconciliation latency, IP claim allocation latency,
     controller restart count, Ironic API response time — all
     available through
     Prometheus metrics.

### Dependencies

- Describe the specific running services the project depends on
  in the cluster.
   - Kubernetes API server, Ironic (with its httpd, dnsmasq,
     and MariaDB/SQLite components), and optionally DHCP/TFTP
     infrastructure for PXE boot. All dependencies except
     Kubernetes can be deployed as containers within the

>      management cluster.

Reviewer: By default also cert-manager is required, although it is possible to bring-your-own certs.

     or the Kubernetes API is temporarily unavailable,
     reconciliation retries automatically. Bare metal hosts
     continue running unaffected; only lifecycle operations are
     paused until connectivity is restored.
- Describe the known failure modes.
   - BMC unreachable (network or credential issues) — surfaces
     as BaremetalHost error condition
   - Firmware bugs causing inspection/provisioning failures —
     documented workarounds per vendor
   - Ironic database corruption — requires database restore
     from backup
   - DHCP/PXE network misconfiguration — host fails to boot IPA
   - Race conditions during concurrent provisioning of large
     batches — mitigated by work queue rate limiting

All failure modes surface as BaremetalHost error conditions
   with descriptive messages and are documented in the
   [troubleshooting guide][troubleshooting].

### Compliance

- What steps does the project take to ensure that all
  third-party code and components have correct and complete
  attribution and license notices?
   - Maintains an Apache 2.0 LICENSE in every repository.
     All third-party Go dependencies are pinned in

>      `go.mod`/`go.sum`, and license attribution is verified

Reviewer: How is it verified? I am not aware of this procedure

   - How are notices managed for third-party code incorporated
     directly into the project's source files?
      - Apache 2.0 headers go in source files. Vendored code
        keeps its original licenses.
   - How are notices retained for unmodified third-party
     components included within the project's repository?
      - The project does not vendor third-party source code;
        dependencies are consumed via Go modules, so no
        aggregated NOTICE file is needed.
   - How are notices for all dependencies obtained at build time
     included in the project's distributed build artifacts
     (e.g. compiled binaries, container images)?
      - SBOM attestation is attached to container images via
        cosign. Go's built-in license tracking is used for
        binary dependencies.

### Security (Day 2)

- Security Hygiene
   - How is the project executing access control?
      - Each controller (BMO, CAPM3, IPAM) runs with its own
        RBAC-scoped ServiceAccount with least-privilege
        permissions. BMC credentials are stored exclusively in
        Kubernetes Secrets and are never exposed in CRD status,

>         logs, or events.

Reviewer: ```suggestion
        logs, or events. All user access is through the Kubernetes API, relying on RBAC for access control.
```

We could possibly also mention BMO <-> Ironic here but I am not sure it is relevant. Ironic is anyway outside of it all.

     features proposed as issues in the metal3-docs repository
     accompanied by [design documents][design-docs]. Proposals
     are discussed in weekly community meetings, annual Metal3
     meetups, and curated by the designated "Roadmap editors"
     team. Roadmap editors are appointed from the maintainer
     pool.
- Describe the target persona or user(s) for the project?
   - The project targets platform engineers and infrastructure
     operators managing bare metal Kubernetes clusters, primarily
     in Telco, edge computing, and on-premise data center
     environments
     ([adopters list][adopters]).
- Explain the primary use case for the project. What additional
  use cases are supported by the project?
   - Declarative lifecycle management of bare metal servers as
     Kubernetes nodes using Cluster API — including provisioning,
     hardware inspection, disk cleaning, and decommissioning —
     via standard BMC protocols (Redfish, IPMI). Additional:
     standalone bare metal management without CAPI via Baremetal
     Operator directly, IP address management for bare metal
     networks.
- Explain which use cases have been identified as unsupported by
  the project.

>    - Metal3 does not manage cloud provider VMs, orchestrate

Reviewer 2: Metal3 does manage non k8s workloads, a BMH or a PPI  are non kubernetes workloads as it is not a computational workload nor purely data kept in etcd,  rather a representation of a non kubernetes concept in k8s space.

Reviewer 2: BMO is a translation layer between the K8s API and whatever non K8s API the provisioner(s) introduce, Metal3 does reach outside the K8s API boundary via BMOs provisioner interface.

     exclusively for bare metal "self-hosted" workflows and does
     not require any commercial cloud provider service.
- Describe the intended types of organizations who would benefit
  from adopting this project.
   - Telco operators (Ericsson, Deutsche Telekom), enterprise
     infrastructure providers (Red Hat, SUSE, Fujitsu), telco
     cloud infrastructure platforms (Sylva, Airship), AI/ML
     infrastructure (Mistral AI), and large-scale on-premise
     environments (Mirantis). Deployments range from
     single-server setups to 2000+ node production
     environments.
- Please describe any completed end user research and link to any
  reports.
   - Adopter feedback is gathered through the CNCF incubation
     adopter interviews (Ericsson, Red Hat, SUSE, Deutsche
     Telekom, and a fifth Telco adopter — see
     [incubation DD][incubation]), direct engagement at community
     meetings, KubeCon presentations, and the
     [Metal3 annual meetup][meetup-notes].

### Usability

- How should the target personas interact with your project?
   - Operators interact entirely through Kubernetes-native APIs
     by creating and managing BaremetalHost, Metal3Machine,

>      Metal3Cluster, IPPool, and IPClaim custom resources using

Reviewer 2: ```suggestion
     Metal3Cluster, IPPool,  IPClaim and other custom resources using
```

     standard tools like kubectl and clusterctl.
- Describe the user experience (UX) and user interface (UI) of
  the project.

>    - Metal3 has no graphical UI; the interface is the Kubernetes

Reviewer 2: There is a live project to integrate with Headlamp, since it is an official LFX project and the mentee has been selected I would mention it as it is WIP: https://mentorship.lfx.linuxfoundation.org/project/dbf009dd-6eed-4c15-b6fc-391c0bc418dc .

So it can be mentioned IMO that there is no Metal3 specific UI but Metal3 and Headlamp community is working together to create a Metal3 headlamp plugin.

     API with CRDs and status conditions. Operators use kubectl (or other

>      Kubernetes API clients)to inspect BaremetalHost status, events, and conditions,

Reviewer 3: ```suggestion
     Kubernetes API clients) to inspect BaremetalHost status, events, and conditions,
```

     which surface the full hardware lifecycle state machine
     (registering → inspecting → available → provisioning →
     provisioned → deprovisioning).
   - The core experience is the BaremetalHost Custom Resource —
     to provision a host, a user writes a YAML manifest with
     BMC credentials and desired state. No graphical dashboard
     or vendor-specific tooling is required.
   - In Cluster API-based workflows, operators interact through
     Cluster API resources (for example Cluster,
     KubeadmControlPlane, and MachineDeployment) while CAPM3
     reconciles the corresponding Metal3 resources
     (Metal3Cluster and Metal3Machine) and drives bare metal
     lifecycle actions.

> - Describe how this project integrates with other projects in a

Reviewer 2: Ironic, Ironic TOR network config tool , Ironic metrics exporter, IPA , cloud init I think these could be all mentioned especially that OpenInfra and CNCF live under the same roof nowadays.

Reviewer 2: Also different CAPI compatible bootstrap providers can be mentioned e.g. ignition bootstrap provider, fake kubeadm provider (used by IBM research at some point to combine Metal3 with IBM mainframes), k3s bootstrap provider (Suse and the Sylva project folks used Metal3 with k3s)

  production environment.
   - Metal3 implements the [Cluster API infrastructure provider
     contract][capi] for bare metal, integrates OpenStack Ironic
     as its provisioning backend, and works with any CNI, CSI,
     and Kubernetes distribution. It is composable with the
     Metal3 IPAM or external IPAM solutions, and integrates with
     observability backends via Prometheus metrics.

### Design

- Explain the design principles and best practices the project
  is following.
   - Kubernetes-native CRDs + controllers; declarative,
     vendor-neutral bare metal management using standard BMC

>      protocols; delegation of hardware provisioning to

Reviewer 2: Provisioner interface and PreprovisioningImageController could be mentioned here with the focus on the fact that the provisioner layer is not hardcoded but extensible via the provisioner interface, so anything that implements the interface could serve as a provisioner.

Reviewer 2: It could be also mentioned that  users can scope the Metal3 stack according to their needs, CAPI and CAPM3 are not mandatory, one could just run with BMO, or consume BMHs with other controllers or use CAPI+CAPM3 with some other BMH consumer in tandem . So technically everything "above" and "below the BMO is modular .

     OpenStack Ironic rather than reimplementing it;
     separation of concerns across independent controllers
     (BMO, CAPM3, IPAM).
- Outline or link to the project's architecture requirements?
  Describe how they differ for Proof of Concept, Development,
  Test and Production environments, as applicable.
   - PoC/dev: [metal3-dev-env][dev-env] provides a virtualized
     environment using QEMU/KVM VMs with virtual BMCs
     (sushy-tools for Redfish, vbmc for IPMI) and a Kind or
     Minikube management cluster.
   - Production: requires BMC network connectivity, Ironic
     deployment with appropriate database backend (MariaDB for
     scale), and HA controller configuration with leader
     election.
   - End-to-end tests run in CI across BMO, CAPM3, and IrSO
     repositories covering install, provision, deprovision,
     and upgrade workflows.
   - Architecture documented at
     [book.metal3.io/architecture][arch].
- Define any specific service dependencies the project relies on
  in the cluster.

>    - Kubernetes API server (with etcd), Ironic deployed via

Reviewer 2: ```suggestion
   - Kubernetes API server (with etcd), Ironic or any other compatible provisioner
```
IrSO is optional so please mention it as an optional dependency if someone would like to configure Ironic via CRs then IrSO is the way but it is not mandatory.

     the [Ironic Standalone Operator][irso] (recommended for

>      production), and a MariaDB or SQLite backend for Ironic

Reviewer 2: Sqlite is not an external dependency it is a file inside the ironic container, I would not mention it as a cluster level dependency, I would mention :

```suggestion
     production), and  MariaDB or other SQL compatible relational database in case the user opts for external database for Ironic
```

     controller-runtime for HA deployments. Ironic can be
     deployed in active-passive HA configurations using the
     [Ironic Standalone Operator][irso]. The BaremetalHost state
     machine is idempotent and resumes correctly after controller
     restarts.
- Describe the project's resource requirements, including CPU,
  Network and Memory.
   - Upstream BMO and CAPM3 manifests do not prescribe fixed
     CPU or memory requests/limits for controller pods.
     Resource usage depends on fleet size, reconciliation
     churn, and provisioning concurrency, so operators should
     baseline from runtime metrics and set requests/limits for
     their environment. The Ironic Python Agent (IPA), which
     runs temporarily on each host during provisioning and
     inspection, is the main per-host resource consumer.
     Ironic itself can be tuned for large-scale deployments.
- Describe the project's storage requirements, including its use
  of ephemeral and/or persistent storage.
   - Storage needs are primarily in Ironic and image-serving
     components, not in BMO/CAPM3 controllers. Ironic can run
     with an ephemeral local SQLite database or with a

>      persistent external MariaDB database. In production,

Reviewer 2: MariaDB is just one example it should be able to run with PostgreSQL or any other SQL compatible relational  DB.

        list): infrastructure and machine APIs (for example
        Metal3Cluster, Metal3Machine, Metal3MachineTemplate,
        Metal3Data and Metal3DataTemplate), host lifecycle and
        inventory APIs (for example BaremetalHost, HardwareData,
        HostFirmwareComponents, HostClaim), and IPAM APIs
        (IPPool, IPClaim, IPAddress). When IRSO is installed, it
        also adds Ironic CRDs. The exact set evolves by release.
        Metal3 does not change Kubernetes core APIs and does not
        add cloud-provider APIs. Controllers call BMC endpoints
        (for example Redfish or IPMI) and Ironic APIs inside the
        deployment.
   - Describe compatibility of any new or changed APIs with API
     servers, including the Kubernetes API server
      - CRDs follow standard Kubernetes structural schema
          validation.
   - Describe versioning of any new or changed APIs, including
     how breaking changes are handled
      - Current released Metal3 APIs are v1beta1. API versions
        follow Kubernetes conventions (v1alpha1, v1beta1).
        Unreleased `main` branch work toward newer Cluster API
        contracts is not reflected here yet. Check the release

>         notes for version history.

Reviewer 3: This is now outdated, we have v1beta2.

- Describe the project's release processes, including major,
  minor and patch releases.
   - Releases happen aligned with Cluster API and Kubernetes
     cycles (~every 4 months) following SemVer, with patch
     releases as needed. Releases are cut from stable branches
     (two supported release lines plus one additional
     CI-tested line), published via GitHub Releases with
     changelogs, signed with cosign, and include SBOMs.
- Describe how the project is installed and initialized, e.g. a
  minimal install with a few lines of code or does it require
  more complex integration and configuration?
   - Installation via `clusterctl init --infrastructure metal3`
     (Cluster API CLI) or kustomize manifests.
    The upstream [quick start][quickstart] walks through a
     complete initial setup for both direct BaremetalHost
    provisioning and Cluster API-based deployments.

>      Ironic is deployed as containers either within the

Reviewer 2: This installation description covers CAPM3, CAPI, IPAM and Ironic but not how IrSO is installed or BMO is installed, sadly we have to mention that they are installed via kustomize or static manifest as we do not provide Helm charts or a cli for these.

      - TLS certificates for Ironic communication can be managed
        via cert-manager or manual rotation. The Ironic
        Standalone Operator supports automated certificate
        management. In standard CAPM3/IPAM/BMO/IRSO
        installations, webhook certificates are issued and
        rotated by cert-manager, which also maintains the
        Secrets mounted by the controller pods.
   - Describe how the project is following and implementing
     secure software supply chain best practices.
      - Released container images are signed with keyless
        cosign and include SPDX SBOM attestations in the
        release workflows. Go module dependencies are pinned in
        `go.sum` and verified by Go module checksum
        validation. Dependabot and periodic OSV scanning are
        used to detect known vulnerabilities. The project
        follows [CNCF SSCP][sscp] best practices.

## Day 1 - Installation and Deployment Phase

### Project Installation and Configuration

- Describe what project installation and configuration look like.
   - Typical bootstrap flow: create a bootstrap Kubernetes
     cluster (or reuse an existing one), install cert-manager,

>      deploy Ironic via IRSO (unless Ironic is already provided

Reviewer 2: IRSO installation has to be mentioned too.

     conversion compatibility for stored objects, but this is
     not equivalent to a full component downgrade
     certification.
- Explain how the project informs users of deprecations and
  removals of features and APIs.
   - Through release notes, the [Metal3 book][book], and
     community meetings. API field deprecations follow the
     Kubernetes deprecation policy with at least one release of
     overlap.
- Explain how the project permits utilization of alpha and beta
  capabilities as part of a rollout.
   - Alpha features are gated behind feature flags in controller
     configuration and are disabled by default. API versions
     follow Kubernetes conventions (v1alpha1, v1beta1).
     Component maturity:

     | Component | Maturity | Version |
    | :---------- | :--------- | :-------- |
      | Baremetal Operator (BMO) | **Stable** | v0.13.x |
      | Cluster API Provider Metal3 (CAPM3) | **Stable** | v1.13.x |
      | IP Address Manager (IPAM) | **Stable** | v1.13.x |
     | Ironic Standalone Operator (IrSO) | **Beta** | v0.x |
     | Ironic Image | **Stable** | Released with BMO |
     | IPA Downloader | **Stable** | Released with BMO |

>      | MariaDB Image | **Stable** | Released independently |

Reviewer 3: Indenting wrong, some 4 some 5.

     host counts, with no fan-out amplification pattern in
     controller logic.
- Describe how the project defines Service Level Objectives
  (SLOs) and Service Level Indicators (SLIs).
   - Exposes Prometheus metrics for reconciliation latency,
     provisioning success rate, and controller error rates that
     adopters can use to define their own SLOs. Formal SLOs are
     not yet defined by the project.
- Describe any operations that will increase in time covered by
  existing SLIs/SLOs.
   - Larger fleet sizes increase total reconciliation loop time.
     Concurrent provisioning operations are bounded by Ironic's
     conductor thread pool.
- Describe the increase in resource usage in any components as a
  result of enabling this project, to include CPU, Memory,
  Storage, Throughput.
   - Pretty minimal footprint per controller (~256Mi memory,
     0.5 CPU). Usage scales linearly with managed
     BaremetalHosts. Ironic database size grows proportionally
     to host count.
- Describe which conditions enabling / using this project would
  result in resource exhaustion of some node resources (PIDs,
  sockets, inodes, etc.)

>    - Extremely large fleet sizes (thousands of hosts) with

Reviewer 2: Extreme is subjective , I would just say "thousands of hosts" and up to an equal number of clusters.

     hosts; larger-scale testing is performed by adopters
     (Ericsson, Red Hat, SUSE) in their production pipelines.
     Ironic scalability has also been tested manually with fake
     IPA agents at 1000-cluster scale.
- Describe the recommended limits of users, requests, system
  resources, etc. and how they were obtained.
   - A single BMO instance comfortably manages hundreds of
     BaremetalHosts; for 1000+ hosts, Ironic tuning (conductor
     workers, database connection pooling) is recommended.
     Limits determined through adopter production experience.
- Describe which resilience pattern the project uses and how,
  including the circuit breaker pattern.
   - Uses the Kubernetes controller-runtime pattern with work
     queues, exponential backoff, rate limiting, and leader
     election. The BaremetalHost state machine is idempotent —
     controllers reconcile from observed state and recover
     automatically from transient failures.

### Observability Requirements

- Describe the signals the project is using or producing,
  including logs, metrics, profiles and traces. Please include
  supported formats, recommended configurations and data
  storage.

>    - Produces Prometheus metrics (reconciliation duration, error

Reviewer 2: In addition to this you could mention that Ironic has it's own node exporter that generates prometheus compatible output. All components running in the cluster generate controller logs and the BMH CR is designed to have extensive status information so that the state of the managed physical machines is visible in detail via the BMH status fields.  Ironic also collects the logs of the IPA and logs it as a pod log so Metal3 even retrives logs from the preprovisioning agents if configured to do so. Inspection could be mentioned that is also kind of a monitoring tool as we can "monitor" the HW composition of the physical machines.

     counts, queue depth) via controller-runtime's built-in
     endpoint on `:8443/metrics`. Controller logs are emitted
     through controller-runtime logging (klog/logr, with zap in
     some components/configurations). Ironic services emit their
     own service logs. Distributed tracing is not enabled by
     default.
- Describe how the project captures audit logging.
   - Relies on Kubernetes audit logging for API operations. The
     controller logs all reconciliation actions. BMC credential
     access is logged but credentials are redacted from all log
     output.
- Describe any dashboards the project uses or implements as well
  as any dashboard requirements.
   - Metal3 does not ship built-in dashboards. Its Prometheus
     metrics can be scraped and visualized with tooling chosen

>      by adopters (for example Grafana, Headlamp, or other

Reviewer 2: I would not group Headlamp with Grafana and other tools together, Headlamp is a K8s UI that has to know the CRs beforehand and needs the WIP Metal3 plugin to function properly and it does not display metrics information by default, Grafana and other dashboard tools are a separate category that indeed can be used for showing monitoring data but then adding "or other Kubernetes observability UIs" could mean tracer tools like Jager and we are not really compatible with them. 

 I would phrase it like "Metal3 is compatible with any dashboard solution that can represent promethes compatible time series data " or something like this.

     Kubernetes observability UIs) to track BareMetalHost state
     distribution, provisioning latency, and error rates.
- Describe how the project surfaces project resource requirements
  for adopters to monitor cloud and infrastructure costs, e.g.
  FinOps
   - BaremetalHost status exposes hardware inventory (CPU, RAM,
     disk, NIC details) discovered during inspection, enabling
     adopters to build asset tracking and capacity planning on
     top of the CRD data.
- Which parameters is the project covering to ensure the health
  of the application/service and its workloads?

>    - Health is tracked across BMH, CAPM3, and IPAM resources:

Reviewer 2: BMH does display BMC health info since 0.13 too.

     reconciliation counts confirm active operation.
- How can someone using this project know that it is working for
  their instance?
   - BaremetalHost resources should show expected state
     transitions (registering → inspecting → available),
     Metal3Machine/Metal3Cluster resources should converge to
     ready states, IPAM objects should show successful claim and
     allocation, reconciliation metrics should be healthy, and
     no persistent errors should appear in controller logs.
- Describe the SLOs (Service Level Objectives) for this project.
   - Not formally defined by the project. Adopters define SLOs
     based on their fleet size and operational requirements.
- What are the SLIs (Service Level Indicators) an operator can
  use to determine the health of the service?
   - Reconciliation success rate, provisioning success rate,
     BaremetalHost time-to-available, Metal3Machine/Cluster
     reconciliation latency, IP claim allocation latency,
     controller restart count, Ironic API response time — all
     available through
     Prometheus metrics.

### Dependencies

- Describe the specific running services the project depends on
  in the cluster.

>    - Kubernetes API server, Ironic (with its httpd, dnsmasq,

Reviewer 2: Ironic or other suitable provisioner, when used with the default Ironic then Ironic also relies on ( apache http server, dnsmasq, ipxe fimware , firmware builder, Ironic TOR switch management components, IPA  etc..)

>      and MariaDB/SQLite components), cert-manager by default

Reviewer 2: plase make sure MariaDB is just an example of external relation DBs, we do not depend specifically on MariaDB.

     for certificate management, and optionally DHCP/TFTP
     infrastructure for PXE boot. Operators can bring their own
     certificates, but cert-manager is the default deployment

>      path. All dependencies except Kubernetes can be deployed

Reviewer 2: Technically true, preprovisioning agent and the disk images has to be either built into the container but commonly they are either downloaded in runtime or attached via volume mounts to the Ironic pod.

Also BMCs are always external to the cluster in production.
